\documentclass{beamer} 
\usetheme{CambridgeUS}

\usepackage[utf8x,utf8]{inputenc} % make weird characters work
\usepackage[english]{babel}

\usepackage{graphicx}
\graphicspath{{Figures/}{./}} % Specifies where to look for included images
\usepackage{color}

\usepackage{mathpazo} % Use the Palatino font by default
\usepackage{amsmath}


\usepackage{verbatim}

\setbeamertemplate{footline}[frame number]{}
\setbeamertemplate{navigation symbols}{}
% \setbeamertemplate{footline}{ }


% -------------------------------------------------

% definišemo korisne komande
\newcommand{\en}[1]{(engl. \textit{#1})}
\newcommand{\lat}[1]{(latin. \textit{#1})}
\newcommand{\keyword}[1]{\textbf{#1}}
\newcommand{\ikeyword}[1]{\textit{\textbf{#1}}}
\newcommand{\tabhead}[1]{\textbf{#1}}
\newcommand{\code}[1]{\texttt{#1}}
\newcommand{\file}[1]{\texttt{\bfseries#1}}
\newcommand{\option}[1]{\texttt{\itshape#1}}


\newcommand{\uniprot}{\textit{UniProt} }
\newcommand{\uniprotkb}{\textit{UniProtKB} }
\newcommand{\swissprot}{\textit{Swiss-Prot} }
\newcommand{\trembl}{\textit{TrEMBL} }

\newcommand{\kw}[1]{\textbf{\textit{KW: #1}}}
\newcommand{\mf}[1]{\textbf{\textit{MF: #1}}}

\newtheorem{definicija}{Definition}

% -------------------------------------------------

% \title{Bioinformatička analiza povezanosti funkcije i neuređenosti proteina}
\title[Protein function and IDP \& IDPr]{ Bioinformatics analysis
of correlation between protein function and intrinsic disorder}
% \subtitle{Master Rad}

\author{
  Goran Vinterhalter$^*1$
  \\ Jovana J Kovačević$^1$
  \\ Gordana Pavlović Lažetić$^1$
}
\institute[Faculty of Mathematics]{
  University of Belgrade, Faculty of Mathematics$^1$
}
\date{June 22, 2018}

\titlegraphic{\includegraphics[scale=0.4]{belbi}}


% This is only inserted into the PDF information catalog. Can be left
% out. 

% If you have a file called "university-logo-filename.xxx", where xxx
% is a graphic format that can be processed by latex or pdflatex,
% resp., then you can add a logo as follows:

\pgfdeclareimage[height=0.5cm]{university-logo}{logo.png}
\logo{\pgfuseimage{university-logo}}


% Delete this, if you do not want the table of contents to pop up at
% the beginning of each subsection:

% \AtBeginSubsection[]
% {
%   \begin{frame}<beamer>{Outline}
%     % \tableofcontents[currentsection,currentsubsection]
%     \tableofcontents[currentsection]
%   \end{frame}
% }

% Let's get started


\AtBeginSection[]
{
  \ifnum \value{framenumber}>1
  \begin{frame}<beamer>
    % \frametitle{Plan}
    \tableofcontents[currentsection]
  \end{frame}
  \else
  \fi
}







\begin{document}

\setbeamertemplate{caption}{\raggedright\insertcaption\par}


\begin{frame}
  \titlepage
\end{frame}

\begin{frame}
\tableofcontents[]
\end{frame}

\section{Inspiration}
\begin{frame}
  \frametitle{Inspiration}

  \begin{block} {Reference work: Xie. et. al 2006}
    \textit{Functional anthology of intrinsic disorder. 1.\\
  \small Biological processes and functions of proteins with long disordered regions}
  \textit{
    \footnotesize
\\Hongbo Xie, Slobodan Vucetic, Lilia M. Iakoucheva, Christopher J. Oldfield,
A. Keith Dunker, Vladimir N. Uversky, and Zoran Obradovic
  }
  % \\ \textit{\footnotesize Xie $H^1$, Vucetic S, Iakoucheva LM, Oldfield CJ, Dunker AK, Uversky VN, Obradovic Z.}
  \end{block}

  \pause
  \begin{itemize}
    \item \swissprot v48 from 2005. ( about 200 000 seq.)
      \pause
    \item UniprotKB Keywords
      \pause
  %   \item \swissprot contains statistically redundant sequences
  %   \item Classtering into protein families  (27 217 families)

    \item PONDR VSL3 disorder predictor
  \end{itemize}

\end{frame}


\section{Method and data}


\subsection{Data}
\begin{frame}
  \frametitle{Used data}

  \begin{itemize}
    \item \textit{CAFA3} dataset (training set from CAFA3 challenge (2016))
      \pause
    \item GO terms
      \pause
    \item \textit{CAFA3} dataset is a subset of \swissprot
      \pause
      \begin{figure}
        \centering 
        \hspace*{-0.5cm} 
        \includegraphics[scale=0.8]{plots/cafa3_pe}
        \vspace*{-0.5cm} 
        \caption{Protein existance}
      \end{figure}

      \pause
    \item We assume the \textit{CAFA3} dataset is not statistically redundant
  \end{itemize}


\end{frame}




\subsection{Dependency between protein length and predicted disorder }

\begin{frame}{Predicting disorder}

  PONDR VSL2b disorder predictor
  \pause

  \begin{definicija}
    \label{pdis_def}
    Protein is labeled as \keyword{putatively disordered} if it contains a
    predicted IDPr at least 40 AA long, otherwise putatively ordered.
  \end{definicija}


  \pause
  {\huge
  But! \\}
  Likelihood of labeling a protein putatively disordered increses by it's length.

\end{frame}


\begin{frame}{}



  \begin{itemize}
    \item
      $S_L$, set of proteins with lengths from interval
      $[L-k, L+k], k=L/10$
      \begin{figure}
        \centering
        \includegraphics[scale=1]{S_L}
        \caption{ $ S_L = \{s_i :  | L -  | s_i | | \le k  \} $ }
      \end{figure}

      \pause

    \item

      The probability a protein of length $L$ is labeled putatively disordered:

      $$ P_L = \dfrac{ \sum_{s_i \in S_L} d(s_i)} {| S_L |} $$
      \[   
        d(s_i) = 
        \begin{cases}
        1 & \text{if  $s_i$ is \keyword{putatively disordered}}\\
        0 & \text{otherwise}
      \end{cases}
    \]

\end{itemize}
\end{frame}


\begin{frame}{}
  \begin{figure}[]
    \centering
    \includegraphics[scale=0.8]{plots/PL_F}
    \caption {
      \footnotesize Dependency between protein length and fraction of putative
      disorder
    }
  \end{figure}
\end{frame}

\begin{frame}{Generating protein sequences}
  \begin{columns}
    \column{0.5\textwidth}
  \begin{itemize} \item random model ( $P_L random$  ) \end{itemize}
    \column{0.5\textwidth}
    \begin{itemize} \item uniform model ( $P_L uniform$ ) \end{itemize}
  \end{columns}

  \begin{figure}[th]
    \centering
    \includegraphics[scale=0.7]{plots/AK_ucestalost}
    \vspace{-0.2cm}
    \caption{ \footnotesize Amino acid fractions in CAFA3 dataset }
  \end{figure}
\end{frame}

\begin{frame}
  \begin{figure}[th]
    \centering
    \includegraphics[scale=0.8]{plots/PL_F_cmp}
    \caption{Comparing $P_L$, $P_L random$ and $P_L uniform$ on CAFA3 dataset}
  \end{figure}
\end{frame}


\subsection{Functions dependency on (dis)order}
\begin{frame}{Correlation between function and (dis)order}

  \begin{itemize}
    \item 
      Let $S_j$ be a set of proteins annotated by function $j$.

      By \keyword{disorder level} (of function $j$) we assume the fractions
      ($F_j$) of putatively disordered proteins inside set $S_j$.

      $$F_j = \dfrac{\sum_{s_i \in S_j} d(s_i)} {|S_j|} $$

      \pause

    \item Null model
      % Nultu hipotezu $Y_j$ za $F_j$  def. preko Bernulijeve slučajne prom. $X_L$
      $$
      Y_j = \dfrac {\sum_{s_i \in S_j} {X_{|s_j|}}}{|S_j|}
      \quad \quad \quad
    X_L : \begin{pmatrix} 0 & 1\\ 1-P_L & P_L \end{pmatrix}
      $$

      \pause

      \begin{itemize}
        \item $p>0.95$: function is \keyword{disorder related}
          (correlated with ...)
        \item $p<0.05$: function is \keyword{order related}
        \item otherwise: nothing can be said about correlation (insufficient stat. sig.)
      \end{itemize}


  \end{itemize}

\end{frame}

% \begin{frame}{Definitions}
%   \begin{definicija}
%     \keyword{Statistički najznačajnije} uređene/neuređene funkcije
%     (GO-termini ili ključne reči) su one koje imaju najmanji/najviši Z-skor.
%   \end{definicija}
% \end{frame}


\section{Data preprocessing}
\subsection{Combining dataset and annotations}
\begin{frame}

  \begin{enumerate}
    \item Merging \swissprot and it's CAFA3 subset
      \pause
      \begin{itemize}
        \item 66 530 valid proteins (out of 66 599 total)
        \item Annotations (term \& kw) we get from \swissprot
        \item Sequences we keep from CAFA3 dataset
      \end{itemize}
      \pause
    \item Annotation grouping for MF terms
  \begin{columns}
    \column{0.25\textwidth}
    \includegraphics[scale=0.7]{grupisanje}
    \column{0.7\textwidth}
    \pause
    1781 MF terms that annotate min. 20 proteins.
    Without this step we wold only get 1146 MF terms.
  \end{columns}
      \pause
    \item
      Mapping MF keywords to MF terms ???
      \pause


  \end{enumerate}

  \begin{table}
    \begin{tabular}{|r|c|c|c|c|c|}
      \hline
      & total & no map. & MF map. & BP map. & CC. map.     \\
      \hline
      MF keywords  & 195    &  20       &  104     & 54      & 11           \\
      \hline
    \end{tabular}
  \end{table}
  
\end{frame}


\subsection{Mapping MF keywords to MF terms}


% \begin{frame}
%   \begin{figure}[!th]
%     \centering
%     \vspace*{-0.2cm} 
%     \includegraphics[scale=0.5]{Figures/plots/kw_dis2go.pdf}
%   \end{figure}
% \end{frame}
%
\begin{frame}
  \begin{figure}[!th]
    \centering
    \vspace*{-0.49cm} 
    \includegraphics[scale=0.45]{Figures/plots/Neuropeptide2go.pdf}
  \end{figure}
\end{frame}


\begin{frame}{Approximated mapping (Annotation similarity)}

  \begin{columns}
    \column{0.7\textwidth}
    \keyword{Jaccard index (Ji}),
      $J(A,B) = \dfrac{|A \cap B|}{|A \cup B|} \in [0, 1] \quad $
    \column{0.3\textwidth}
      \includegraphics[scale=1]{Ji}
  \end{columns}

  \pause

  \begin{itemize}
    \item Consider the whole \swissprot (but really CAFA3 dataset is enough)
      \pause
    \item No MF annotations grouping (smaller Ji, or wrong) 
      \pause
  \item expecting between 60 and 85\% "correct" mapping
      \pause
  \item We produce 64 mappings with $Ji>0.1$

  \end{itemize}
\end{frame}


\begin{frame}

  \hspace*{-0.4cm}
  \scriptsize
  \begin{tabular}{|p{4.2cm}|c|c|c|p{4.3cm}|}
    \hline
    \bf \textit{MF keywords } & \bf \# kw & \bf Ji & \bf \# go & \bf \textit{MF terms} \\
    \hline
    \hline
    Dermonecrotic toxin                & 148   & 0.96  & 142   & phospholipase D activity \\ \hline
    \keyword{Ribosomal protein}        & 49054 & 0.91  & 48096 & structural constituent of ribosome \\ \hline
    Complement sys. impairing toxin & 160   & 0.81  & 142   & phospholipase D activity \\ \hline
    Hemagglutinin                      & 397   & 0.75  & 299   & host cell surface receptor binding \\ \hline
    Mutator protein                    & 255   & 0.75  & 288   & damaged DNA binding \\ \hline
    Antifreeze protein                 & 10    & 0.7   & 7     & ice binding \\ \hline
    Light-harvesting polypeptide       & 90    & 0.68  & 61    & bacteriochlorophyll binding \\ \hline
    Cyclin                             & 197   & 0.61  & 124   & cyclin-dependent protein ... \\ \hline
    Defensin                           & 55    & 0.55  & 32    & CCR6 chemokine receptor binding \\ \hline
    \keyword{Ribonucleoprotein}        & 50698 & 0.54  & 28317 & rRNA binding \\ \hline
    Neurotoxin                         & 2734  & 0.53  & 4145  & toxin activity \\ \hline
    Photoprotein                       & 40    & 0.48  & 19    & alkanal monooxygenase ... \\ \hline
    \keyword{Endorphin}                & 48    & 0.45  & 32    & opioid peptide activity \\ \hline
    Protein synthesis inhibitor        & 150   & 0.43  & 67    & rRNA N-glycosylase activity \\ \hline
    \keyword{Neuropeptide}             & 561   & 0.42  & 267   & neuropeptide hormone activity \\ \hline
    Signal transduction inhibitor      & 157   & 0.38  & 158   & GTPase activator activity \\ \hline
    Mitogen                            & 282   & 0.37  & 284   & growth factor activity \\ \hline
    Repressor                          & 8177  & 0.33  & 7798  & DNA binding \\ \hline
    ... & ... & ... & ... &  ...\\
    \hline
    \keyword{Vasoactive}                            & 243   & 0.17  & 489   & hormone activity \\ \hline
    \keyword{Chromatin regulator}                   & 1939  & 0.12  & 861   & chromatin binding \\ \hline
    \keyword{Developmental protein}                 & 6285  & 0.12  & 2464  & sequence-specific DNA binding \\ \hline
  \end{tabular}

\end{frame}




\section{Results}


\begin{frame}{Results}

  \begin{table}[htpb]
    \centering
    \begin{tabular}{|r|c|c|c|}
      \hline
      & \small reference & \multicolumn{2}{c|}{results} \\
      & Xie2007 kw &  MF kw  & MF terms  \\
      \hline
      total ($\# prot\ge20$)         & 143  & 186    & 1781          \\
      $p<0.05$ (order rel.)          & 37   & 53     & 699           \\
      $p>0.95$ (disorder rel.)       & 51   & 44     & 616           \\
       insufficient stat. sig.         & 55   & 89     & 576           \\
      \hline
    \end{tabular}
  \caption{}
  \end{table}

  \pause

  Two types of results:
  \begin{itemize}
    \item Tabular comparison with keywords
    \item GO subgraph
  \end{itemize}

\end{frame}

\subsection{Tabular comparison}


\begin{frame}
  \begin{figure}[th]
    \vspace*{-0.8cm}
    \hspace*{-0.45cm}
    \includegraphics[ scale=0.23]{Figures/plots/order_cmp.pdf}
  \end{figure}
\end{frame}

\begin{frame}
  \begin{figure}[th]
    \vspace*{-0.8cm}
    \hspace*{-0.45cm}
    \includegraphics[ scale=0.23]{Figures/plots/disorder_cmp.pdf}
  \end{figure}
\end{frame}


\subsection{GO subgraph}

\begin{frame}
  \begin{figure}
    \vspace{-0.7cm}
    \includegraphics[ scale=0.75]{Figures/plots/disorder_example.pdf}
  \end{figure}
\end{frame}


\subsection{Level of disorder and p-value}
\begin{frame}
  \begin{figure}
    \centering
    \vspace{-0.5cm}
    \includegraphics[scale=0.55]{Figures/plots/avg_dis2p.pdf}
    \caption {
      Level of disorder (\textit{avg\_dis}) and p-value (1781 MF terms)
    }
  \end{figure}
\end{frame}


\begin{frame}
  \begin{center}
    \Huge Thank you
  \end{center}
\end{frame}



\end{document}
