{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from lark import Lark\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def apply_meta_rules(gramatika):\n",
    "    gramatika = re.sub(r\"([a-z_]+)-TVP\",      r' \"\\1:\" UNQUOTED_STRING -> \\1',   gramatika)\n",
    "    gramatika = re.sub(r\"([a-z_]+)-BT\",       r' \"\\1:\" ( TRUE | FALSE ) -> \\1',  gramatika)\n",
    "    gramatika = re.sub(r\"([a-z_]+)-Tag (.*)\", r' \"\\1:\" \\2 -> \\1',                gramatika)\n",
    "    return gramatika"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 208 ms, sys: 0 ns, total: 208 ms\n",
      "Wall time: 206 ms\n"
     ]
    }
   ],
   "source": [
    "parser = Lark(apply_meta_rules(r\"\"\"\n",
    "    start : not_important* term_frame+ typedef?\n",
    "    typedef : \"[Typedef]\" /(.|\\n)+/\n",
    "    not_important : /.+/? _NL\n",
    "    \n",
    "    TRUE: \"true\"\n",
    "    FALSE: \"false\"\n",
    "       \n",
    "    class_id : id\n",
    "    rel_id : id\n",
    "    instance_id : id\n",
    "    ?id : URL_AS_ID | PREFIXED_ID | UNPREFIXED_ID\n",
    "    URL_AS_ID  : (\"https:\" | \"http:\")  /(?:[^\\s,\\}\\]\\\\\\\\]|\\\\\\\\.)+/ \n",
    "    PREFIXED_ID :  /[^\\s:=]+:\\s*(?:[^\\s,\\}\\]\\\\\\\\]|\\\\\\\\.)+/             \n",
    "    UNPREFIXED_ID: /(?:[^\\s:,=\\}\\]\\\\\\\\]|\\\\\\\\.)+/\n",
    " \n",
    " \n",
    "        UNQUOTED_STRING: /.+/\n",
    "\n",
    "// line termination\n",
    "    eol :  qualifier_block? HIDDEN_COMMENT? _NL\n",
    "    HIDDEN_COMMENT : \"!\" /[^\\n]*/\n",
    "    ?qualifier_block : \"{\" qualifier_list? \"}\"\n",
    "    qualifier_list :  qualifier (\",\" qualifier)*\n",
    "    qualifier : rel_id  \"=\"  QUOTED_STRING                 // requiers that ID dosn't allow '='\n",
    "    \n",
    "    \n",
    "    xref_list : \"[\" (xref_no_comma (\",\" xref_no_comma)*)? \"]\"\n",
    "    xref: id QUOTED_STRING?               // instead of 'id' it cold be /\\S+/ ???, (does it have to escape ] ?)\n",
    "    xref_no_comma: id QUOTED_STRING ?\n",
    "    \n",
    "        term_frame : \"[Term]\" _NL \"id:\" class_id eol (term_frame_clause eol )*\n",
    "        \n",
    "        ?term_frame_clause : is_anonymous-BT\n",
    "                           | name-TVP\n",
    "                           | namespace-Tag OBO_NAMESPACE\n",
    "                           | alt_id-Tag id\n",
    "                           | def-Tag QUOTED_STRING xref_list\n",
    "                           | comment-TVP\n",
    "                           | subset-Tag id                                           // Subset-ID  (def in header) \n",
    "                           | synonym-Tag QUOTED_STRING SYNONYM_SCOPE NAME? xref_list   // synonym-ID (def in header)\n",
    "                           | xref-Tag xref\n",
    "                           | builtin-BT\n",
    "                           | property_value-Tag RELATION_ID   /.+/        // TODO\n",
    "                           | is_a-Tag class_id\n",
    "                           \n",
    "                           | intersection_of-Tag  /.+/        // TODO: RElATION_ID? class_id\n",
    "                           \n",
    "                           | union_of-Tag class_id\n",
    "                           | equivalent_to-Tag class_id\n",
    "                           | disjoint_from-Tag class_id\n",
    "                           | relationship-Tag RELATION_ID class_id\n",
    "                           | is_obsolete-BT\n",
    "                           | replaced_by-Tag class_id\n",
    "                           | consider-Tag id\n",
    "                           | created_by-Tag /.+/       // Person-ID\n",
    "                           | creation_date-Tag /.+/       // ISO-8601-DateTime\n",
    "                           //| /[^:]+:.+/          -> anything\n",
    "\n",
    "                           \n",
    "                            \n",
    "        OBO_NAMESPACE : \"cellular_component\" | \"biological_process\" | \"molecular_function\"\n",
    "        \n",
    "        SYNONYM_SCOPE : \"EXACT\" | \"BROAD\" | \"NARROW\" | \"RELATED\"\n",
    "        \n",
    "        RELATION_ID : /\\S+/\n",
    "        \n",
    "        %import common.ESCAPED_STRING -> QUOTED_STRING\n",
    "        %import common.CNAME -> NAME\n",
    "        %import common.NEWLINE -> _NL\n",
    "        %import common.WS_INLINE\n",
    "        %ignore WS_INLINE\n",
    "        \n",
    "        \n",
    "\n",
    "    \"\"\"), parser=\"lalr\", lexer=\"contextual\")\n",
    "\n",
    "text = r\"\"\"\n",
    "[Term]\n",
    "id: GO:32\n",
    "def: \"hello\" [wikipedia:nesto[lepo\\]tako]\n",
    "\"\"\"\n",
    "\n",
    "with open(\"../data/go_1.obo\", \"r\") as obo_file:\n",
    "    #tree = parser.parse(text)\n",
    "    tree = parser.parse(obo_file.read())\n",
    "    pass\n",
    "    \n",
    "# def showTree(t, ind=\"\"):\n",
    "#     global tree\n",
    "#     if not t:\n",
    "#         return\n",
    "    \n",
    "#     print(ind, end='')\n",
    "#     if type(t) == type(tree):\n",
    "#         print(t.data)\n",
    "#         for childe in t.children:\n",
    "#             showTree(childe, ind+\" \"*2)\n",
    "#     else:\n",
    "#         print(repr(t))\n",
    "        \n",
    "# showTree(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from lark import Transformer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Token(PREFIXED_ID, 'GO:0000001')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'mitochondrion inheritance')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'biological_process')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"The distribution of mitochondria, including the mitochondrial genome, into daughter cells after mitosis or meiosis, mediated by interactions between mitochondria and the cytoskeleton.\"'), ['GOC:mcc', 'PMID:10873824', 'PMID:11389764']])\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"mitochondrial inheritance\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0048308')]])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0048311')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000002')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'mitochondrial genome maintenance')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'biological_process')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"The maintenance of the structure and integrity of the mitochondrial genome; includes replication and segregation of the mitochondrial chromosome.\"'), ['GOC:ai', 'GOC:vw']])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0007005')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000003')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'reproduction')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'biological_process')])\n",
      "None\n",
      "Tree(alt_id, [Token(PREFIXED_ID, 'GO:0019952')])\n",
      "None\n",
      "Tree(alt_id, [Token(PREFIXED_ID, 'GO:0050876')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"The production of new individuals that contain some portion of genetic material inherited from one or more parent organisms.\"'), ['GOC:go_curators', 'GOC:isa_complete', 'GOC:jl', 'ISBN:0198506732']])\n",
      "None\n",
      "goslim_chembl\n",
      "None\n",
      "goslim_generic\n",
      "None\n",
      "goslim_pir\n",
      "None\n",
      "goslim_plant\n",
      "None\n",
      "gosubset_prok\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"reproductive physiological process\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(xref, [Tree(xref, [Token(PREFIXED_ID, 'Wikipedia:Reproduction')])])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0008150')]])\n",
      "None\n",
      "Tree(disjoint_from, [[Token(PREFIXED_ID, 'GO:0044848')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000005')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'obsolete ribosomal chaperone activity')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'molecular_function')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"OBSOLETE. Assists in the correct assembly of ribosomes or ribosomal subunits in vivo, but is not a component of the assembled ribosome when performing its normal biological function.\"'), ['GOC:jl', 'PMID:12150913']])\n",
      "None\n",
      "Tree(comment, [Token(UNQUOTED_STRING, 'This term was made obsolete because it refers to a class of gene products and a biological process rather than a molecular function.')])\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"ribosomal chaperone activity\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(is_obsolete, [Token(TRUE, 'true')])\n",
      "None\n",
      "Tree(consider, [Token(PREFIXED_ID, 'GO:0042254')])\n",
      "None\n",
      "Tree(consider, [Token(PREFIXED_ID, 'GO:0044183')])\n",
      "None\n",
      "Tree(consider, [Token(PREFIXED_ID, 'GO:0051082')])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000006')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'high-affinity zinc uptake transmembrane transporter activity')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'molecular_function')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"Enables the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+(out) = Zn2+(in), probably powered by proton motive force. In high-affinity transport the transporter is able to bind the solute even if it is only present at very low concentrations.\"'), ['TC:2.A.5.1.1']])\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"high affinity zinc uptake transmembrane transporter activity\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0005385')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000007')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'low-affinity zinc ion transmembrane transporter activity')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'molecular_function')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"Catalysis of the transfer of a solute or solutes from one side of a membrane to the other according to the reaction: Zn2+ = Zn2+, probably powered by proton motive force. In low-affinity transport the transporter is able to bind the solute only if it is present at very high concentrations.\"'), ['GOC:mtg_transport', 'ISBN:0815340729']])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0005385')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000008')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'obsolete thioredoxin')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'molecular_function')])\n",
      "None\n",
      "Tree(alt_id, [Token(PREFIXED_ID, 'GO:0000013')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"OBSOLETE. A small disulfide-containing redox protein that serves as a general protein disulfide oxidoreductase. Interacts with a broad range of proteins by a redox mechanism, based on the reversible oxidation of 2 cysteine thiol groups to a disulfide, accompanied by the transfer of 2 electrons and 2 protons. The net result is the covalent interconversion of a disulfide and a dithiol.\"'), ['GOC:kd']])\n",
      "None\n",
      "Tree(comment, [Token(UNQUOTED_STRING, 'This term was made obsolete because it represents gene products.')])\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"thioredoxin\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(is_obsolete, [Token(TRUE, 'true')])\n",
      "None\n",
      "Tree(consider, [Token(PREFIXED_ID, 'GO:0003756')])\n",
      "None\n",
      "Tree(consider, [Token(PREFIXED_ID, 'GO:0015036')])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0000009')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'alpha-1,6-mannosyltransferase activity')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'molecular_function')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"Catalysis of the transfer of a mannose residue to an oligosaccharide, forming an alpha-(1->6) linkage.\"'), ['GOC:mcc', 'PMID:2644248']])\n",
      "None\n",
      "Tree(term_frame_clause, [Token(QUOTED_STRING, '\"1,6-alpha-mannosyltransferase activity\"'), Token(SYNONYM_SCOPE, 'EXACT'), []])\n",
      "None\n",
      "Tree(xref, [Tree(xref, [Token(PREFIXED_ID, 'EC:2.4.1')])])\n",
      "None\n",
      "Tree(xref, [Tree(xref, [Token(PREFIXED_ID, 'Reactome:REACT_22295'), Token(QUOTED_STRING, '\"Addition of a third mannose to the N-glycan precursor by Alg2, Saccharomyces cerevisiae\"')])])\n",
      "None\n",
      "Tree(xref, [Tree(xref, [Token(PREFIXED_ID, 'Reactome:REACT_22383'), Token(QUOTED_STRING, '\"Addition of a third mannose to the N-glycan precursor by ALG2, Homo sapiens\"')])])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0000030')]])\n",
      "None\n",
      "==============================\n",
      "[Token(PREFIXED_ID, 'GO:0099616')]\n",
      "None\n",
      "Tree(name, [Token(UNQUOTED_STRING, 'extrinsic component of matrix side of mitochondrial inner membrane')])\n",
      "None\n",
      "Tree(namespace, [Token(OBO_NAMESPACE, 'cellular_component')])\n",
      "None\n",
      "Tree(def, [Token(QUOTED_STRING, '\"The component of the matrix side of the mitochondrial inner membrane consisting of gene products and protein complexes that are loosely bound to one of its surfaces, but not integrated into the hydrophobic region.\\\\\"\\\\n\"'), ['GOC:dos']])\n",
      "None\n",
      "Tree(is_a, [[Token(PREFIXED_ID, 'GO:0031314')]])\n",
      "None\n",
      "GO:0019898 ! extrinsic component of membrane\n",
      "None\n",
      "part_of GO:0099617 ! matrix side of mitochondrial inner membrane\n",
      "None\n",
      "Tree(relationship, [Token(RELATION_ID, 'part_of'), [Token(PREFIXED_ID, 'GO:0099617')]])\n",
      "None\n",
      "==============================\n"
     ]
    }
   ],
   "source": [
    "class OBOTransformer(Transformer):  \n",
    "    \n",
    "    TRUE = lambda self, _: True\n",
    "    FALSE = lambda self, _: False\n",
    "    \n",
    "    eol = lambda self, _: None\n",
    "    \n",
    "    def id(self, item): return item[0].value  # str\n",
    "    def class_id(self, item): return item  # str\n",
    "    def rel_id(self, item): return item[0].value  # str\n",
    "    def instance_id(self, item): return item[0].value  # str\n",
    "    def relation_id(self, item): return item[0].value  # str\n",
    "\n",
    "    def xref_no_comma(self, items): return items[0].value # TODO\n",
    "    xref_list = list\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "\n",
    "for x in OBOTransformer().transform(tree).children:\n",
    "    for y in x.children:\n",
    "        print y\n",
    "    print \"=\"*30"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:master]",
   "language": "python",
   "name": "conda-env-master-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3.0
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}